### PR TITLE
Use `RoomVersion` objects

### PR DESCRIPTION
**Pull request by richvdh**
Various refactors to use `RoomVersion` objects instead of room version identifieers.

Should be reviewable commit-by-commit

### CODE REVIEW

File: synapse/handlers/federation.py [hunk 1]
```diff
@@ -897,9 +897,9 @@ async def on_make_leave_request(
             )
             raise SynapseError(403, "User not from origin", Codes.FORBIDDEN)
 
-        room_version = await self.store.get_room_version_id(room_id)
-        builder = self.event_builder_factory.new(
-            room_version,
+        room_version_obj = await self.store.get_room_version(room_id)
```
Comment (clokep) [inline synapse/handlers/federation.py:900]: I think we likely could have kept the variable names as `room_version`?

File: synapse/events/builder.py [hunk 2]
```diff
@@ -197,24 +195,6 @@ def __init__(self, hs: "HomeServer"):
...
-        Returns:
-            EventBuilder
-        """
-        v = KNOWN_ROOM_VERSIONS.get(room_version)
-        if not v:
-            # this can happen if support is withdrawn for a room version
-            raise UnsupportedRoomVersionError()
-        return self.for_room_version(v, key_values)
-
     def for_room_version(
```
Comment (clokep) [inline synapse/events/builder.py:198]: I wonder if we should rename this to `new(...)` now?

**Review by clokep — APPROVED** [review]
LGTM! The comments are both at your discretion!

File: synapse/handlers/federation.py (hunk 1 above)
Comment (richvdh) [reply synapse/handlers/federation.py:900]: yeah we could; however renaming them made it easier for me to make sure I was updating all the places they were referenced, and I think it does no harm to have renamed them.

File: synapse/events/builder.py (hunk 2 above)
Comment (richvdh) [reply synapse/events/builder.py:198]: I kinda like having the verbose name. It makes grepping more effective, for a start.

**Pull request merged**